### PR TITLE
Add support  of 'decimal' option to Series.to_csv and Dataframe.to_csv

### LINKED ISSUE #781: Handle European decimal formats in to_csv
Counterpart to issue #584

### PR DESCRIPTION
**Pull request by bertrandhaut**
closes #781

The 'decimal' option exists for read_csv method but not yet in 'to_csv' methods.
The lack of this option is particulary painful when we _have to_ work with Excel with European regional settings. 

This modification add this option to both Series.to_csv and Dataframe.to_csv.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
@jreback Can this be revisited? 

Is there something that needs to be changed by @bertrandhaut (apart from adding a vbench)

**Comment by shoyer** [comment]
I think this needs a test?

**Comment by bertrandhaut** [comment]
As suggested, I've added a test for this new option.

File: pandas/core/frame.py [hunk 1]
```diff
@@ -1126,6 +1126,8 @@ def to_csv(self, path_or_buf=None, sep=",", na_rep='', float_format=None,
         date_format : string, default None
             Format string for datetime objects
         cols : kwarg only alias of columns [deprecated]
+        decimal: string, default '.'
+            Character recognized as decimal separator. E.g. use ‘,’ for European data
```
Comment (jorisvandenbossche) [inline pandas/core/frame.py]: Can you use standard single quotes here? `'` instead of `‘`

File: pandas/core/internals.py [hunk 2]
```diff
@@ -1171,10 +1171,22 @@ def to_native_types(self, slicer=None, na_rep='', float_format=None,
         values = np.array(values, dtype=object)
         mask = isnull(values)
         values[mask] = na_rep
-        if float_format:
+        
+        
+        if float_format and decimal != '.':
+            formater = lambda v : (float_format % v).replace('.',decimal,1)
+        elif decimal != '.':
+            formater = lambda v : ('%g' % v).replace('.',decimal,1)
```
Comment (jorisvandenbossche) [inline pandas/core/internals.py]: Use `formatter` instead of `formater` as the name of this function?

File: pandas/core/series.py [hunk 3]
```diff
@@ -2267,6 +2267,8 @@ def to_csv(self, path, index=True, sep=",", na_rep='',
             non-ascii, for python versions prior to 3
         date_format: string, default None
             Format string for datetime objects.
+        decimal: string, default '.'
+            Character recognized as decimal separator. E.g. use ‘,’ for European data
```
Comment (jorisvandenbossche) [inline pandas/core/series.py]: same remark here

File: pandas/tests/test_format.py [hunk 4]
```diff
@@ -2343,7 +2343,15 @@ def test_csv_to_string(self):
...
-
+        
+    def test_to_csv_decimal(self):
+        df = DataFrame({'col1' : [1], 'col2' : ['a'], 'col3' : [10.1] })
+        
+        expected_default = ',col1,col2,col3\n0,1,a,10.1\n'
+        self.assertEqual(df.to_csv(), expected_default)
+
+        expected_european_excel = ';col1;col2;col3\n0;1;a;10,1\n'
+        self.assertEqual(df.to_csv(decimal=',',sep=';'), expected_european_excel)
```
Comment (jorisvandenbossche) [inline pandas/tests/test_format.py]: Can you add a test for the case that also `float_format` is specified?

**Comment by bertrandhaut** [comment]
Joris' comments taken into account

**Comment by jorisvandenbossche** [comment]
@jreback Is this good to go? (not really familiar with this)

@bertrandhaut Before merging, we ask to squash the commits into one. Can you do that (https://github.com/pydata/pandas/wiki/Using-Git#fetch-and-then-rebase-interactively-to-squash-reword-fix-otherwise-change-some-commits, if you have questions, just ask!)

File: pandas/tests/test_format.py [hunk 5]
```diff
@@ -2343,7 +2343,21 @@ def test_csv_to_string(self):
         df = DataFrame({'col' : [1,2]})
         expected = ',col\n0,1\n1,2\n'
         self.assertEqual(df.to_csv(), expected)
-
+        
+    def test_to_csv_decimal(self):
+        df = DataFrame({'col1' : [1], 'col2' : ['a'], 'col3' : [10.1] })
```
Comment (jreback) [inline pandas/tests/test_format.py]: add the issue number as a comment here

**Comment by jreback** [comment]
@bertrandhaut can you do a vbench run on the csv's to make sure that perf is unchanged, see here:https://github.com/pydata/pandas/wiki/Performance-Testing

File: pandas/tests/test_format.py (hunk 5 above)
Comment (bertrandhaut) [inline pandas/tests/test_format.py]: done

Comment (bertrandhaut) [inline pandas/tests/test_format.py]: I've tried to squash the commits as described but I faced the following error:
$git rebase pandas/master
fatal: Needed a single revision
invalid upstream pandas/master

Any idea ?

**Comment by bertrandhaut** [comment]
For the vbench I tried but without success. For information I've only python3 on my computer and it seems that vbench is still not ported to python3.

File: pandas/tests/test_format.py (hunk 5 above)
Comment (shoyer) [inline pandas/tests/test_format.py]: Try this:
```
git checkout master
git pull pandas/master master
git checkout your-branch
git rebase master
```

**Comment by jorisvandenbossche** [comment]
About the vbench, I think that is correct that is not yet ported to python 3.

File: pandas/tests/test_format.py (hunk 5 above)
Comment (bertrandhaut) [inline pandas/tests/test_format.py]: There should be something special that I do to create my local version...

When I do the command "git checkout master" in my to level directory I get
    error: pathspec 'master' did not match any file(s) known to git.

If I do it in the "pandas" directory (the one with the setup.py file) I get:
   Already on 'master'
   Your branch is up-to-date with 'origin/master'.
but then the second command git pull pandas/master master lead to
    fatal: 'pandas/master' does not appear to be a git repository
    fatal : Could not read from remote repository


By the way, in my repository, I don't think I've made any branch. I've done
everything in master. Maybe was that not a good idea ?



On Tue, Mar 3, 2015 at 9:31 AM, Stephan Hoyer <notifications@github.com>
wrote:

> In pandas/tests/test_format.py
> <https://github.com/pydata/pandas/pull/8448#discussion_r25670074>:
>
> > @@ -2343,7 +2343,21 @@ def test_csv_to_string(self):
> >          df = DataFrame({'col' : [1,2]})
> >          expected = ',col\n0,1\n1,2\n'
> >          self.assertEqual(df.to_csv(), expected)
> > -
> > +
> > +    def test_to_csv_decimal(self):
> > +        df = DataFrame({'col1' : [1], 'col2' : ['a'], 'col3' : [10.1] })
>
> Try this:
>
> git checkout master
> git pull pandas/master master
> git checkout your-branch
> git rebase master
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydata/pandas/pull/8448/files#r25670074>.
>



-- 
Bertrand Haut

**Comment by jreback** [comment]
@bertrandhaut can you rebase and squash to a single commit. looks ok otherwise to me.

**Comment by jorisvandenbossche** [comment]
@bertrandhaut You were indeed working on master. In the future, better to always first make a branch if you are going to work on a certain feature/fix. But for this PR, it is OK.

Given you are working on master, this should work to rebase/squash:

```
git fetch upstream
git rebase -i upstream/master
```

**Comment by jreback** [comment]
merged via https://github.com/pydata/pandas/commit/671c4b3c52a02ee3736752b2645d08762c0fedaf

thanks!

**Pull request closed without merging**